lib/core/src/display/world.rs

```diff
@@ -1,143 +1,207 @@
+            data.self_reference = Some(world_ref3);
+        });
+        world_ref
     }
-
-    pub fn started(&self) -> bool {
-        self.data.borrow().started
+    /// Create new uninitialized world instance. You should rather not need to
+    /// call this function directly.
+    pub fn new_uninitialized() -> Self {
+        let workspaces       = default();
+        let logger           = Logger::new("world");
+        let workspace_logger = logger.sub("workspace_dirty");
+        let workspace_dirty  = WorkspaceDirty::new(workspace_logger,());
+        let event_loop       = EventLoop::new();
+        let update_handle    = default();
+        let self_reference   = default();
+        Self {workspaces,workspace_dirty,logger,event_loop,update_handle
+             ,self_reference}
     }
-
-    pub fn start(&self) {
-        if !self.started() {
-            self.data.borrow_mut().started = true;
-            Self::request_callback(&self.on_every_frame);
+    /// Add new workspace and get its ID.
+    pub fn add_workspace(&mut self, name: &str) -> WorkspaceID {
+        let logger = &self.logger;
+        let dirty  = &self.workspace_dirty;
+        self.workspaces.insert_with_ix(|ix| {
+            group!(logger, format!("Adding workspace {} ({}).", ix, name), {
+                let on_change     = workspace_on_change(dirty.clone_rc(),ix);
+                let wspace_logger = logger.sub(ix.to_string());
+                Workspace::new(name,wspace_logger,on_change).unwrap() // FIXME
+            })
+        })   
+    }
+    /// Dispose the workspace by the provided ID. In case of invalid ID, a 
+    /// warning will be emitted.
+    pub fn drop_workspace(&mut self, id: WorkspaceID) {
+        let logger = &self.logger;
+        let item   = self.workspaces.remove(id);
+        match item {
+            None => logger.warning("Trying to delete non-existing workspace."),
+            Some(item) => group!(logger, "Dropping workspace {}.", id, {
+                let _destruct_it_here = item;
+            }),
         }
     }
-
-    pub fn stop(&self) {
-        self.data.borrow_mut().started = false;
+    /// Run the provided callback on every frame. Returns a `CallbackHandle`, 
+    /// which when dropped will cancel the callback. If you want the function
+    /// to run forever, you can use the `forget` method in the handle. 
+    pub fn on_frame<F:FnMut(&mut World)+'static>
+    (&mut self, mut callback: F) -> CallbackHandle { 
+        let this = self.self_reference.as_ref().unwrap().clone_rc();
+        let func = move || callback(&mut this.borrow_mut());
+        self.event_loop.add_callback(func)
+    }
+    /// Check dirty flags and update the state accordingly.
+    pub fn update(&mut self) {
+        if self.workspace_dirty.check() {
+            group!(self.logger, "Updating.", {
+                self.workspace_dirty.unset();
+                self.workspaces.iter_mut().for_each(|t| t.update());
```

Comment:
What is purpose of different dirty flag types, if we're updating all workspaces anyway?

Reply:
This fragment is not finished. We will soon be considering each of them separately.